src/main/host/syscall/handler/epoll.rs

```diff
@@ -0,0 +1,423 @@
+            if let Some(abs_timeout) = cond.timeout() {
+                if Worker::current_time().unwrap() >= abs_timeout {
+                    log::trace!("No events are ready on epoll {epfd} and the timeout expired");
+                    return Ok(0);
+                }
+            }
+        }
+
+        // If there's a signal pending, this syscall will be interrupted.
+        if ctx.objs.thread.unblocked_signal_pending(
+            ctx.objs.process,
+            &ctx.objs.host.shim_shmem_lock_borrow().unwrap(),
+        ) {
+            return Err(SyscallError::new_interrupted(false));
+        }
+
+        // Convert timeout to an EmulatedTime.
+        let Ok(abs_timeout_opt) = timeout
+            .map(|x| Worker::current_time().unwrap().checked_add(x).ok_or(()))
+            .transpose()
+        else {
+            log::trace!("Epoll wait with invalid timeout {timeout:?} (too large)");
+            return Err(Errno::EINVAL.into());
+        };
+
+        log::trace!("No events are ready on epoll {epfd} and we need to block");
+
+        // Block on epoll status; an epoll descriptor is readable when it has events.
+        let mut rv = SyscallError::new_blocked(
+            File::Epoll(Arc::clone(epoll)),
+            FileState::READABLE,
+            /* restartable= */ false,
+        );
+
+        // Set timeout, if provided.
+        if abs_timeout_opt.is_some() {
+            rv.blocked_condition().unwrap().set_timeout(abs_timeout_opt);
+        }
+
+        Err(rv)
+    }
+}
+
+fn timeout_arg_to_maybe_simtime(
+    timeout: std::ffi::c_int,
+) -> Result<Option<SimulationTime>, SyscallError> {
+    // epoll_wait(2): "Specifying a timeout of -1 causes epoll_wait() to block indefinitely"
+    let timeout = (timeout >= 0).then_some(timeout);
+
+    if let Some(timeout) = timeout {
+        // a non-negative c_int should always convert to a u64
+        let timeout = timeout.try_into().unwrap();
+        let timeout = SimulationTime::try_from_millis(timeout).ok_or(Errno::EINVAL)?;
+        Ok(Some(timeout))
+    } else {
+        Ok(None)
+    }
+}
+
+fn epoll_max_events_upper_bound() -> i32 {
+    // https://github.com/torvalds/linux/blob/master/fs/eventpoll.c#L2299
+    let ep_max_events = i32::MAX;
+    let ep_ev_size: i32 = std::mem::size_of::<linux_api::epoll::epoll_event>()
+        .try_into()
+        .unwrap_or(i32::MAX);
+    ep_max_events.saturating_div(ep_ev_size)
+}
+
+fn write_events_to_ptr(
+    mem: &mut MemoryManager,
+    ready: Vec<(EpollEvents, u64)>,
+    events_ptr: ForeignPtr<linux_api::epoll::epoll_event>,
+) -> Result<(), SyscallError> {
+    let events_ptr = ForeignArrayPtr::new(events_ptr, ready.len());
+    let mut mem_ref = mem.memory_ref_mut(events_ptr)?;
+
+    for (i, (ev, data)) in ready.iter().enumerate() {
+        let plugin_ev = &mut mem_ref.deref_mut()[i];
+        plugin_ev.events = ev.bits();
+        plugin_ev.data = *data;
+    }
```

Comment:
nit: what about something like the following to avoid the array indexing and possible panic?

```rust
for ((ev, data), plugin_ev) in ready.iter().zip(mem_ref.deref_mut().iter_mut())
```